src/Workspaces/Core/Portable/Storage/SQLite/SQLitePersistentStorage.Accessor.cs

```diff
@@ -87,31 +71,32 @@ public Task<Stream> ReadStreamAsync(TKey key, Checksum checksum, CancellationTok
 
                     if (haveDataId)
                     {
-                        // Ensure all pending document writes to this name are flushed to the DB so that 
-                        // we can find them below.
-                        await FlushPendingWritesAsync(key, cancellationToken).ConfigureAwait(false);
-
-                        try
+                        using (var pooledConnection = Storage.GetPooledConnection())
                         {
-                            using (var pooledConnection = Storage.GetPooledConnection())
+                            // First, try to see if there was a write to this key in 
+                            // our in-memory db.
+                            var result = ReadBlob(
+                                pooledConnection.Connection, WriteCacheDBName,
+                                dataId, columnName, checksumOpt, cancellationToken);
+                            if (result != null)
                             {
-                                // Lookup the row from the DocumentData table corresponding to our dataId.
-                                return ReadBlob(
-                                    pooledConnection.Connection, dataId, columnName,
-                                    checksumOpt, cancellationToken);
+                                return result;
                             }
-                        }
-                        catch (Exception ex)
-                        {
-                            StorageDatabaseLogger.LogException(ex);
+
+                            // Wasn't in the in-memory write-cache.  Check the full on-disk file.
+
+                            // Lookup the row from the DocumentData table corresponding to our dataId.
+                            return ReadBlob(
+                                pooledConnection.Connection, MainDBName,
+                                dataId, columnName, checksumOpt, cancellationToken);
```

Comment:
note: i believe this code is race-free (and i will add a commetn as to why).

Basically, let's say that we have anotehr thread trying to do the "flush write db to disk" work.  That work is happening in a transaction.  So the question is how the two reads here interplay with that transaction.  A read will either see the work pre or post the transaction.  So, if we read from the cache and we don't see anything.  That's either because there wasn't anything in the cache, or because we've seen the cache *after* the transaction where all the data was copied from the cache to the db.

In the latter case, the next read from the actual db will see that.

Reply:
Tossing out another idea:  We could write a single compound query that would check both tables (both the write cache and the real db).  I wonder if that approach can be used to get any more performance. Note, however, that it would require a re-think of the race conditions problem.

Reply:
> Tossing out another idea: We could write a single compound query that would check both tables (both the write cache and the real db). 

Ooh.  That would be really interesting to try out.  Especially if we turn off WAL and thus have no race in the "flush from memory to disk" transaction.